dali/operators/decoder/video/video_decoder_base.h

```diff
@@ -15,121 +15,437 @@
+            sample_shape[1] = frames_decoders_[s]->Height();
+            sample_shape[2] = frames_decoders_[s]->Width();
+            sample_shape[3] = frames_decoders_[s]->Channels();
+          },
+          -s);
+    }
+    thread_pool.RunAll();
 
-    int64_t f = 0;
-    // Work until:
-    //    (a) There are no more frames, or
-    //    (b) Sufficient number of frames has been decoded.
-    for (; f < num_frames && frames_decoder.NextFrameIdx() != -1; f++) {
-      frames_decoder.ReadNextFrame(output_data + f * frame_size);
-    }
-    assert(f <= num_frames);
-    bool full_sequence_decoded = f == num_frames;
-    // If there's an insufficient number of frames, pad if requested.
-    for (; f < num_frames && pad_value.has_value(); f++) {
-      kernels::copy<storage_backend_for_copy_kernel, storage_backend_for_copy_kernel>(
-              output_data + f * frame_size, pad_value->raw_data(), frame_size,
-              std::is_same_v<storage_backend_for_copy_kernel, StorageGPU> ? *stream : 0);
-    }
-    return full_sequence_decoded;
+    output_desc.resize(1);
+    output_desc[0].shape = std::move(out_shape);
+    output_desc[0].type = DALI_UINT8;
+    return true;
   }
 
+  void RunImpl(Workspace &ws) override {
+    auto &output = ws.Output<OutBackend>(0);
+    const auto &input = ws.Input<InBackend>(0);
+    int batch_size = input.num_samples();
 
-  /**
-   * @brief Decode sample with index `idx` to `output` tensor.
-   */
-  void DecodeSample(SampleView<OutBackend> output, int64_t idx) {
-    int64_t num_frames = output.shape()[0];
-    DecodeFrames(output, idx, num_frames);
+    // Decode samples in parallel
+    ThreadPool &thread_pool = GetThreadPool(ws);
+    for (int s = 0; s < batch_size; ++s) {
+      thread_pool.AddWork(
+          [&, s](int tid) {
+            int64_t num_frames = output[s].shape()[0];  // It was calculated in SetupImpl
+            int orig_num_frames = frames_decoders_[s]->NumFrames();
+            auto &ctx = ctx_[tid];
+            ctx.frames_decoder = frames_decoders_[s].get();
+            ctx.stream = ws.has_stream() ? ws.stream() : 0;
+            ctx.fill_value = fill_value_;
+
+            if (boundary_type_ == boundary::BoundaryType::CONSTANT) {
+              int nfill_values = fill_value_.size();
+              DALI_ENFORCE(nfill_values == 1 || nfill_values == ctx.frames_decoder->Channels(),
+                           make_string("Constant padding requires either a single fill value or a "
+                                       "fill value per channel. "
+                                       "Got ",
+                                       nfill_values, " fill values for a video with ",
+                                       ctx.frames_decoder->Channels(), " channels."));
+            }
+
+            if (frames_.HasValue()) {
+              for (int64_t frame_num = 0; frame_num < num_frames; frame_num++) {
+                DALI_ENFORCE(frames_[s].data[frame_num] < orig_num_frames,
```

Comment:
So we deliberately don't want to pad but error, right?

Reply:
when providing explicit frame indices, yes

Reply:
Giving it one more thought, it is trivial to support it, so if the user really wants that, let it be